protocol/synthetix/subgraph/mainnet/handleDelegationUpdated.ts

```diff
@@ -9,19 +14,50 @@ export function handleDelegationUpdated(event: DelegationUpdated): void {
     .concat('-')
     .concat(event.params.collateralType.toHex());
   let position = Position.load(id);
+
+  let vault = Vault.load(
+    event.params.poolId.toString().concat('-').concat(event.params.collateralType.toHex())
+  );
+  if (vault === null) {
+    vault = new Vault(
+      event.params.poolId.toString().concat('-').concat(event.params.collateralType.toHex())
+    );
+    vault.created_at = event.block.timestamp;
+    vault.created_at_block = event.block.number;
+    vault.collateral_amount = event.params.amount.toBigDecimal();
+    vault.collateral_type = event.params.collateralType.toHex();
+    vault.pool = event.params.poolId.toString();
+  } else {
+    if (position) {
+      const isIncreasing = event.params.amount.toBigDecimal().gt(position.collateral_amount);
+      if (isIncreasing) {
+        log.info('INCREASING {} {}', [
+          event.params.amount.toString(),
+          position.collateral_amount.toString(),
+        ]);
+        vault.collateral_amount = vault.collateral_amount.plus(
+          event.params.amount.toBigDecimal().minus(position.collateral_amount)
+        );
+      } else {
+        vault.collateral_amount = vault.collateral_amount.plus(event.params.amount.toBigDecimal());
```

Comment:
when we modify collateral we actually set value, not "add" collateral.
So the vault value needs to increase by difference between (`event.params.amount` and previous collateral amount).

To simplify logic and reduce ifs I suggest setting "previous amount" to 0 if position does not exist and then all the calculations will be identical.

```js
if (!vault) {
 // create new vault *without* collateral_amount
  previous_vault_amount = 0
} else {
  previous_vault_amount = vault.collateral_amount
}

if (!position) {
 // create new position *without* collateral_amount
 previous_position_amount = 0
} else {
  previous_position_amount = position.colateral_amount 
}

// now fill all the values that are common for update/create

amount_delta = event.params.amount - previousCollateral

vault.collateral_amount = previous_vault_amount + amount_delta
position.collateral_amount = event.params.amount 

```

nested if/else are now so complicated it is pretty much impossible to understand what is going on.

Reply:
"my version" is not tested and just a short summary of the idea.
Please remove 3 levels of `if`s, comments are not enough to understand what is going on there. Logic is too complex.
DelegationUpdated event has actual new amount of what is delegated, not the delta. So `position.collateral_amount = event.params.amount` should be correct in this case. If tests are failing then tests need to be fixed.

See the source: https://github.com/Synthetixio/synthetix-v3/blob/main/protocol/synthetix/contracts/interfaces/IVaultModule.sol#L30

Reply:
gotcha, updated the tests